### PR TITLE
Adding histograms to tensorboard

### PR DESCRIPTION
**Pull request by d4vidbiertmpl**
Hey, 

as this is a small PR and I already implemented this feature for myself I did not open an issue.

Description:
I changed EventStorage and TensorboardXWriter such that histograms (for monitoring weights or gradients) can be saved to tensorboard.

To make this implementation as efficient as possible I calculate the histogram statistics for the add_histogram_raw function of the PyTorch SummaryWriter (instead of saving weight matrices). In general, I tried to match the information flow to that of the put_image function.

In my personal implementation I extended the EventStorage and TensorboardXWriter, but I thought that this might be a feature that Detectron could offer out of the box.

A possible example usage could be a Hook:
```
class HistogramHook(HookBase):
    def after_step(self):
        storage = self.trainer.storage
        hist_period = self.trainer.cfg.HIST_PERIOD

        if hist_period > 0 and storage.iter % hist_period == 0:
            self.add_histogram(self.trainer.model, storage)

    def add_histogram(self, model, storage):
        for i, (tag, value) in enumerate(model.named_parameters()):
            # E.g. add histograms for parameters of the Mask R-CNN mask head
            if tag.split(".")[1] == "mask_head":
                storage.put_histogram(tag, value.data.cpu().numpy())
```
Feedback is very much appreciated! :)

Best,
David

### CODE REVIEW

File: detectron2/utils/events.py [hunk 1]
```diff
@@ -295,6 +302,42 @@ def put_scalars(self, *, smoothing_hint=True, **kwargs):
...
             self.put_scalar(k, v, smoothing_hint=smoothing_hint)
 
+    def put_histogram(self, hist_name, hist_tensor, bins=1000):
+        """
+        Create an histogram of a numpy array, calculate its parameters and add them
+        to self `_histograms`.
+
+        Args:
+            hist_name (str): The name of the histogram to put into tensorboard.
+            hist_tensor (numpy.array): An `uint8` or `float` array
```
Comment (alekseynp) [inline detectron2/utils/events.py:312]: Is this a tensor or an array?

Comment (d4vidbiertmpl) [reply detectron2/utils/events.py:312]: This is a NumPy array of the shape of the given parameter/gradient.

Possible usage could be:
```
storage.put_histogram(hist_name, value.data.cpu().numpy())
```
where value is a PyTorch tensor.

Comment (alekseynp) [reply detectron2/utils/events.py:312]: Maybe if you implemented this with `torch.histc` the whole thing could be tensors?

Comment (d4vidbiertmpl) [reply detectron2/utils/events.py:312]: Good point! I will look into it :)

**Comment by d4vidbiertmpl** [comment]
I have now implemented the same functionality only using PyTorch functions. For calculating ```hist_edges``` I followed the implementation of ```np.histogram()``` to make sure the behavior is the same.

File: detectron2/utils/events.py (hunk 1 above)
Comment (d4vidbiertmpl) [reply detectron2/utils/events.py]: I have now implemented the same functionality using only PyTorch functions. For calculating ```hist_edges``` I followed the implementation of ```np.histogram()``` to make sure the behavior is the same.

**Comment by alekseynp** [comment]
For the record I am currently using this PR and so far so good.

**Comment by shida666** [comment]
Thank you for your contribution！i copy your code to replace my events.py ,but my tensorboard (http://localhost:6007/)has not the histogram of PR.
2020-06-18 12-06-28屏幕截图  can you help  me?please
@d4vidbiertmpl @alekseynp

**Comment by shida666** [comment]
can you show the histogram you got in tensorboard? thanks!

**Pull request closed without merging**